### DISCUSSION TITLE
Sporadic 401 when trying to access /external-auths

### DISCUSSION THREAD

**Discussion by kisamoto** [pocketbase/pocketbase#5119] [Q&A]
Having a strange issue that is hard to reproduce.

I'm using a custom OAuth2 provider for sign-ins only.

Sometimes users start experiencing a 401 when trying to retrieve `external-auths`, it shows up in the PocketBase logs like this:

```
data.url	/api/collections/users/records/sxrs0w3wzors3sx/external-auths
data.error	The request requires admin or record authorization token to be set.
data.details	N/A
data.auth	guest
data.status	401
```

If I force a log out (i.e. `authStore.clear()`) it seems to work again.

My initial hunch is some form of caching of credentials or perhaps the `authStore` is invalid.

Any ideas about how I could fix this? Do I need to implement a check that manually performs an oauth refresh in the frontend if the authStore is invalid?

**Comment by ganigeorgiev** [comment]
Without some **minimal** reproducible steps there is no much I can help with.

I suspect that the issue is client-side related. If you use the JS SDK in a JS SSR context make sure that the SDK instance is unique for each request (aka. if you are experiencing "sporadic 401" errors most likely you are reusing the same `pb` instance between requests from different users; see the [JS SSR example](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration)).

**Reply by kisamoto** [reply]
Yeah I'm trying to create a minimal reproduction but can't even guarantee reproducing it in current deployment.

I'm not using SSR though so I don't _think_ there's any way it could be shared between requests.

**Reply by ganigeorgiev** [reply]
Are you sure that the user auth token is not just expired?

Note that the token doesn't refresh on its own and you have to call `authRefresh` while still being valid.

On second read, I also don't understand what do you mean by _"I'm using a custom OAuth2 provider for sign-ins only"_. How is this provider registered?

**Reply by kisamoto** [reply]
It's most likely just expired. I'm not sure why sometimes it did work and other times not. Do you have any recommendation about where the refresh logic should be put into the app? Is there any reason this couldn't also be baked into the JS SDK so that it checks for token validity before performing a request? 

Sorry it's not clear, I'm just trying to say I'm using a non-social provider (Ory cloud) registered as `oidc` in PocketBase through the admin UI.

**Reply by ganigeorgiev** [reply]
> Do you have any recommendation about where the refresh logic should be put into the app?

Without details about your app and how it operates I'm not sure what to recommend. For SPA I usually invoke `authRefresh` on each page load/refresh (eg. https://github.com/presentator/presentator/blob/master/ui/src/pb.js#L140 _the example is a little more complicated in Presentator because the app has to maintain auths to 2 different collections at the same time_).

> Is there any reason this couldn't also be baked into the JS SDK so that it checks for token validity before performing a request?

I don't understand what do you mean by this. Even if the SDK knows what requests require token or not the end result will be the same - a `ClientResponseError`.

In any case I'll stop engage as I don't think I can help further without some reproducible steps.

**Reply by kisamoto** [reply]
Updating here as I've been poking around.

I was briefly able to have a failing request in one window and, in a private window, a successful request. Comparing the request headers and data, he only difference in request is the `Authorization` header however both are valid (not expired - see the Payload data copied from both below).

[Looking at where the error is thrown](https://github.com/pocketbase/pocketbase/blob/master/apis/middlewares.go#L193), it's unable to find the record in the context. I can see there is a record in the database for given ID (manually checking via SQLite and looking in the `users` table).

My current hypothesis is that where [the record is added to the context](https://github.com/pocketbase/pocketbase/blob/master/apis/middlewares.go#L242) it only adds it if there is no error. If there is a validation error then it doesn't log anything, just doesn't add the record to the context. The record may not be added [for a few reasons](https://github.com/pocketbase/pocketbase/blob/master/daos/record.go#L414) but I'm wondering if it's something to do with the verification with the ParseJWT. This would fail if somehow the tokens got out of sync.

I'll try and set up a local debugger to verify but I was wondering if you have any insights as to why tokens may get out of sync or why verification would be failing? 

Successful request:
```
{
  "collectionId": "_pb_users_auth_",
  "exp": 1720292614,
  "id": "bjtekwf8mnaprwq",
  "type": "authRecord"
}
```

Failing request:
```
{
  "collectionId": "_pb_users_auth_",
  "exp": 1719587361,
  "id": "bjtekwf8mnaprwq",
  "type": "authRecord"
}
```

**Reply by ganigeorgiev** [reply]
The token verification could fail if the secret verification fails. This could happen for example if you are using non-auth token (eg. a password reset) or the user has changed their password (on each password reset all previously issued user tokens are invalidated).
